spikeinterface/curation/remove_duplicated_spikes.py

```diff
@@ -0,0 +1,76 @@
+from typing import Optional
+import numpy as np
+from spikeinterface.core import BaseSorting, BaseSortingSegment
+from spikeinterface.core.core_tools import define_function_from_class
+
+
+class RemoveDuplicatedSpikesSorting(BaseSorting):
+	"""
+	Class to remove duplicated spikes from the spike trains.
+	Spikes are considered duplicated if they are less than x
+	ms appart where x is the censored period.
+
+	Parameters
+	----------
+	sorting: BaseSorting
+		The parent sorting.
+	censored_period_ms: float
+		The censored period to consider 2 spikes to be duplicated (in ms).
+	"""
+
+	def __init__(self, sorting: BaseSorting, censored_period_ms: float = 0.3) -> None:
+		super().__init__(sorting.get_sampling_frequency(), sorting.unit_ids)
+		censored_period = int(round(censored_period_ms * 1e-3 * sorting.get_sampling_frequency()))
+
+		for segment in sorting._sorting_segments:
+			self.add_sorting_segment(RemoveDuplicatedSpikesSortingSegment(segment, censored_period, sorting.unit_ids))
+
+
+class RemoveDuplicatedSpikesSortingSegment(BaseSortingSegment):
+
+	def __init__(self, parent_segment: BaseSortingSegment, censored_period: int, unit_ids) -> None:
+		super().__init__()
+		self._parent_segment = parent_segment
+		self._duplicated_spikes = {}
+
+		for unit_id in unit_ids:
+			self._duplicated_spikes[unit_id] = self._find_duplicated_spikes(parent_segment.get_unit_spike_train(unit_id), censored_period)
+
+	@staticmethod
+	def _find_duplicated_spikes(spike_train: np.ndarray, censored_period: int, seed: int = 186472189):
+		"""
+		To ensure no bias is created, a list of duplicated spikes is kept, and one spike is removed
+		iteratively until there are no more duplicated spikes.
+		"""
+
+		rand_state = np.random.get_state()	# Need to store the old state to not seed globally.
+		np.random.seed(seed)				# Need to seed to have the same result for parallelization.
+
+		indices_of_duplicates = np.where(np.diff(spike_train) <= censored_period)[0]
+		indices_of_duplicates = np.unique(np.concatenate((indices_of_duplicates, indices_of_duplicates + 1)))
+		mask = np.ones(len(indices_of_duplicates), dtype=bool)
+
+		while not np.all(np.diff(spike_train[indices_of_duplicates][mask]) > censored_period):
+			index = np.random.randint(low=0, high=len(mask))
+			mask[index] = False
+
+		np.random.set_state(rand_state)
+
+		return indices_of_duplicates[~mask]
+
+	def get_unit_spike_train(self, unit_id, start_frame: Optional[int] = None, end_frame: Optional[int] = None) -> np.ndarray:
+		spike_train = self._parent_segment.get_unit_spike_train(unit_id)
+		spike_train = np.delete(spike_train, self._duplicated_spikes[unit_id])
+
+		if start_frame == None:
+			start_frame = 0
+		if end_frame == None:
+			end_frame = spike_train[-1]
+
+		start = np.searchsorted(spike_train, start_frame, side="left")
+		end   = np.searchsorted(spike_train, start_frame, side="right")
```

Comment:
You should have a margin with the censored_period for this.

Reply:
All the duplicated spikes have been removed at this stage (even the one not between start_frame and end_frame), so I don't see how a margin is relevant?